### DISCUSSION TITLE
Shared StringBuilder for incremental source generator?

### DISCUSSION THREAD

**Discussion by weifenluo** [dotnet/roslyn#74440] [Q&A] [answered]
What is the threading model of incremental source generator? Can I use the same instance of internal `StringBuilder` for one source generator when generating multiple output files? If the source generator is invoked synchronized, this can have some performance gains.

**Comment by CyrusNajmabadi** [comment, marked as answer]
> What is the threading model of incremental source generator? 

Everything should be assumed to be free-threaded and concurrent. 

>  Can I use the same instance of internal StringBuilder for one source generator when generating multiple output files?

No. 

> If the source generator is invoked synchronized, this can have some performance gains.

Expect that everything is concurrent.  Shared state is effectively verboten unless you are very careful to ensure it is absolutely safe.
